apm-agent-core/src/main/java/co/elastic/apm/agent/report/serialize/DslJsonSerializer.java

```diff
@@ -857,25 +857,29 @@ private void serializeDestination(Destination destination) {
         if (destination.hasContent()) {
             writeFieldName("destination");
             jw.writeByte(OBJECT_START);
+            boolean isNeedAddComma = false;
             if (destination.getAddress().length() > 0) {
                 writeField("address", destination.getAddress());
+                isNeedAddComma = true;
             }
             if (destination.getPort() > 0) {
-                writeField("port", destination.getPort());
+                writeLastField("port", destination.getPort());
+                isNeedAddComma = true;
             }
-            serializeService(destination.getService());
+            serializeService(isNeedAddComma, destination.getService());
             jw.writeByte(OBJECT_END);
             jw.writeByte(COMMA);
         }
     }
 
-    private void serializeService(Destination.Service service) {
+    private void serializeService(boolean isNeedAddComma, Destination.Service service) {
         if (service.hasContent()) {
+            if (isNeedAddComma) {
+                jw.writeByte(COMMA);
+            }
             writeFieldName("service");
             jw.writeByte(OBJECT_START);
-            writeField("name", service.getName());
```

Comment:
We cannot remove these fields in the serialized JSON as APM Server requires them to be set as per the intake API spec. 

https://github.com/elastic/apm-agent-java/blob/ae3c602115183b8dc8be5ae8e27910ad4955d9df/apm-agent-core/src/test/resources/apm-server-schema/current/span.json#L129-L133

This spec https://github.com/elastic/apm/pull/436 (still a draft at the time of writing) defines how to automatically set these fields.

Reply:
fixed. @felixbarny should I delete all withName, withType calls?
![image](https://user-images.githubusercontent.com/48118512/116603485-ca35d180-a94e-11eb-9e44-1fba9e9d9096.png)

Reply:
Let's do that in a follow-up PR once https://github.com/elastic/apm/pull/436 is merged. That follow-up PR can then also handle the automatic setting of `context.destination.service.resource`, as described in the linked spec PR draft.